### PR TITLE
Changes for EOS-22210 Implement Health Monitor Evaluator

### PR DESCRIPTION
**Pull request by ArchanaLimaye**
## Problem Statement
<pre>
  <code>
  Story Ref (if any):
    Add MonitorRulesManager class with functionality evaluate, add_rule, remove_rule
  </code>
</pre>
## Unit testing on RPM done
<pre>
  <code>
  No
  </code>
</pre>
## Problem Description
<pre>
  <code>
    Add MonitorRulesManager class with functionality evaluate, add_rule, remove_rule
  </code>
</pre>
## Solution
<pre>
  <code>
    Added MonitorRulesManager class with functionality evaluate, add_rule, remove_rule
  </code>
</pre>
## Unit Test Cases
<pre>
  <code>
    Testing done on a deployed VM; by calling the MonitorRulesManager functions directly.
    Verification of expected behavior done by checking the output in consul and by means of debug logs / print messages

    Following cases tested
   add_rule
      - Rule added for a resource/event . Verified the addition   ex: add_rule("node:fru:fan", "failed", "publish")
      - Rule added for same resource/different event. Verified that same consul KV gets updated   ex: add_rule("node:fru:fan", "failed", "ha")
      - Same rule added again. Verified that warning log is seen and no change seen on consul ex: add_rule("node:fru:fan", "failed", "publish")

  remove_rule
     - Added a rule having multiple actions  . Ex 'cortx/ha/v1/action/node:fru:disk/failed': '["ha", "publish"]'
     - Removed this rule for one of those actions. Verified that the removed action is not seen in consul KV for that rule . Ex remove_rule("node:fru:fan", "failed", "publish")
     - Removed this same rule again. Verified that wanring log is displayed ex: remove_rule("node:fru:fan", "failed", "publish")
     - Removed the same rule for the second action. Verified that since the value for the rule is now empty, the rule itself is deleted from consul . Ex remove_rule("node:fru:fan", "failed", "ha")
     - Removed a non existing rule. Verified that warning log is displayed

  evaluate
    - created rules by using add_rule
    - created a health object such that resource_type, event_type match to this rule
      confirmed that evaluate returns the correct values matching this rule
    - created a health object that will not matching any rules 
      confirmed that evaluate returns empty list in this case 

  </code>
</pre>

### CODE REVIEW

**Comment by ArchanaLimaye** [comment]
Points to note:
1. add_rule/remove_rule  : class definition changed to 
  add_rule(self, resource: str, event: HEALTH_STATUSES , action: HEALTH_MON_ACTIONS)
  remove_rule(self, resource: str, event: HEALTH_STATUSES , action: HEALTH_MON_ACTIONS)

2. In add_rule, as discussed;  validation for resource, event, action parameters 
   is expected to be done at call point. so no checks are added

**Comment by indrajitzagade** [comment]
@ArchanaLimaye  Please rebase this branch with fault_tolerance branch.

**Comment by ArchanaLimaye** [comment]
@indrajitzagade rebased to latest branch

File: ha/core/health_monitor/const.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+# PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License along
+# with this program. If not, see <https://www.gnu.org/licenses/>. For any questions
+# about this software or licensing, please email opensource@seagate.com or
+# cortx-questions@seagate.com.
+
+from enum import Enum
+
+class HEALTH_MON_ACTIONS(Enum):
```
Comment (ajay-paratmandali) [inline ha/core/health_monitor/const.py:20]: Update Enum class inheritance
https://github.com/Seagate/cortx-ha/blob/fault_tolerance/ha/core/event_manager/const.py
as per above example

File: ha/core/health_monitor/monitor_rules_manager.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            val = json.loads(val)
+        return val
+
+    def add_rule(self, resource: str, event: HEALTH_STATUSES , action: HEALTH_MON_ACTIONS):
+
+        key = f"{HEALTH_MON_KEYS.ACT_RULE.value}/{resource}/{event}"
+        val = []
+
+        Log.info(f"Adding rule for key: {key} ,value: {action}")
+
```
Comment (ajay-paratmandali) [inline ha/core/health_monitor/monitor_rules_manager.py:50]:
```
        if self._confstore.key_exists(key):
            kv = self._confstore.get(key)
            k, val = kv.popitem()
            val = json.loads(val)
```
Repetitive code

```
_get_val(key)
```
return None if key not exists

**Review by ajay-paratmandali — CHANGES_REQUESTED** [review]

File: ha/core/health_monitor/monitor_rules_manager.py (hunk 2 above)
Comment (ArchanaLimaye) [reply ha/core/health_monitor/monitor_rules_manager.py:100]: Done

File: ha/core/health_monitor/const.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+# Foundation, either version 3 of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+# PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License along
+# with this program. If not, see <https://www.gnu.org/licenses/>. For any questions
+# about this software or licensing, please email opensource@seagate.com or
+# cortx-questions@seagate.com.
```
[the remark this answers is not in the document]
Comment (ArchanaLimaye) [reply ha/core/health_monitor/const.py:16]: Done

File: ha/core/health_monitor/monitor_rules_manager.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        Returns:
+            list: actions configured for the rule
+        """
+        val = []
+        key = self._prepare_key(event.resource_type, event.event_type)
+
+        Log.debug(f"Evaluating rule for {key}")
+        kv = self._get_val(key)
+        if kv:
+            print(kv)
```
Comment (ajaysrivas) [inline ha/core/health_monitor/monitor_rules_manager.py:77]: print?

File: ha/core/health_monitor/monitor_rules_manager.py (hunk 2 above)
Comment (ajay-paratmandali) [reply ha/core/health_monitor/monitor_rules_manager.py:100]: k, val = kv.popitem()
            val = json.loads(val)

This also repetitive can be moved to method

File: ha/core/health_monitor/monitor_rules_manager.py (hunk 4 above)
Comment (ArchanaLimaye) [reply ha/core/health_monitor/monitor_rules_manager.py:77]: removed

Comment (ajaysrivas) [reply ha/core/health_monitor/monitor_rules_manager.py:77]: I still see it.

File: ha/core/health_monitor/monitor_rules_manager.py (hunk 2 above)
Comment (ArchanaLimaye) [reply ha/core/health_monitor/monitor_rules_manager.py:114]: done

File: ha/core/health_monitor/monitor_rules_manager.py (hunk 4 above)
Comment (ArchanaLimaye) [reply ha/core/health_monitor/monitor_rules_manager.py]: can you please check now

File: ha/core/health_monitor/monitor_rules_manager.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+class MonitorRulesManager:
+
+    def __init__(self):
+
+        self._confstore = ConfigManager.get_confstore()
+
+    def evaluate(self, event: HealthEvent):
+
+        val = []
+        key = f"{HEALTH_MON_KEYS.ACT_RULE.value}/{event.resource_type}/{event.event_type}"
```
[the remark this answers is not in the document]
Comment (ArchanaLimaye) [reply ha/core/health_monitor/monitor_rules_manager.py]: done

**Review by indrajitzagade — APPROVED** [review]

**Review by ajay-paratmandali — APPROVED** [review]

**Pull request merged**